plaso/parsers/setupapi.py

```diff
@@ -145,47 +182,78 @@ def ParseRecord(self, parser_mediator, key, structure):
     Raises:
       ParseError: when the structure type is unknown.
     """
-    if key != 'logline':
-      raise errors.ParseError(
-          'Unable to parse record, unknown structure: {0:s}'.format(key))
+    if key == 'ignorable_line':
+      return
+
+    if key == 'section_header':
+      self._last_entry_type = self._GetValueFromStructure(
+          structure, 'entry_type')
+      return
+
+    if key == 'section_start':
+      time_structure = self._GetValueFromStructure(structure, 'start_time')
+      start_time = self._BuildDatetime(time_structure)
+      if not start_time:
+        parser_mediator.ProduceExtractionWarning(
+            'invalid date time value: {0!s}'.format(time_structure))
+        return
+
+      event_data = SetupapiLogEventData()
+      event_data.entry_type = self._last_entry_type
 
-    self._ParseRecordLogline(parser_mediator, structure)
+      event = time_events.DateTimeValuesEvent(
+          start_time, definitions.TIME_DESCRIPTION_START,
+          time_zone=parser_mediator.timezone)
+
+      # Create event for the start of the setupapi section
+      parser_mediator.ProduceEventWithEventData(event, event_data)
+      return
 
-  def VerifyStructure(self, parser_mediator, lines):
+    if key == 'section_end':
+      time_structure = self._GetValueFromStructure(structure, 'end_time')
+      end_time = self._BuildDatetime(time_structure)
+      if not end_time:
+        parser_mediator.ProduceExtractionWarning(
+            'invalid date time value: {0!s}'.format(time_structure))
+      self._last_end_time = end_time
+      return
+
+    if key == 'section_end_exit_status':
+      exit_status = self._GetValueFromStructure(
+          structure, 'exit_status')
+      if self._last_end_time:
```

Comment:
Please add a comment that explains why exit time is stored internally first and than later used

Reply:
Done